src/foam/nanos/auth/FilterBySpidService.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+   package: 'foam.nanos.auth',
+   name: 'FilterBySpidService',
+   extends: 'foam.nanos.auth.ProxyUserLocatorService',
+   flags: ['java'],
+
+  documentation: `Restrict login to the url that matches the spid of the user.`,
+
+
+  javaImports: [
+    'foam.dao.DAO',
+    'foam.nanos.auth.AuthenticationException',
+    'foam.nanos.auth.User',
+    'foam.nanos.theme.Theme',
+
+    'static foam.mlang.MLang.*'
+  ],
+
+  properties: [
+    {
+      class: 'String',
+      name: 'superSpid',
+      documentation: 'Set spid of a theme that is accessible to all users',
+      value: "",
+    }
+  ],
+
+  methods: [
+    {
+      name: 'getUser',
+      documentation: 'Helper logic function to reduce code duplication',
+      type: 'User',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'identifier',
+          type: 'String'
+        },
+        {
+          name: 'password',
+          class: 'String'
+        }
+      ],
+      javaCode: `
+        DAO userDAO = (DAO) getX().get("localUserDAO");
+
+        userDAO = userDAO
+          .where(
+          AND(
+            OR(
+              EQ(User.EMAIL, identifier.toLowerCase()),
+              EQ(User.USER_NAME, identifier)
+            ),
+            OR(
+              EQ(User.SPID, ((Theme) x.get("theme")).getSpid()),
+              EQ(User.SPID, getSuperSpid())
+            )
+          ));
+
+        x = x.put("localUserDAO", userDAO);
```

Comment:
I don't think we want to do this but rather let delegates be responsible for user queries. to be discussed on the zoom

Reply:
maybe we do

Reply:
so i think this specific decorator, since its called "FilterBySpid", should only care for the spid. It only should be adding spid predicate on the userDAO